File: Assets/LearnXLua/Scripts/Lesson6_Event.cs
using System;
using UnityEngine;
using XLua;

public class Lesson6_Event : MonoBehaviour
{
    public LuaInit luaInit;
    
    // C# 事件
    public static event Action<string> OnCSharpEvent;
    
    // Lua 回调
    private Action<string> luaCallback;
    
    void Start()
    {
        luaInit.luaEnv.DoString("require 'Lesson6_Event'");
        
        // 获取Lua回调函数
        luaCallback = luaInit.luaEnv.Global.Get<Action<string>>("on_lua_callback");
        
        // 注册Lua到C#的事件监听
        OnCSharpEvent += (message) => {
            Debug.Log($"C# received from Lua: {message}");
        };
    }
    
    [ContextMenu("1. Lua触发事件-Lua接收")]
    void TestLuaToLua()
    {
        luaInit.luaEnv.Global.Get<Action>("test_lua_to_lua")?.Invoke();
    }
    
    [ContextMenu("2. C#触发事件-C#接收")]
    void TestCSharpToCSharp()
    {
        Debug.Log("C# triggering event for C#");
        OnCSharpEvent?.Invoke("C# to C# message");
    }
    
    [ContextMenu("3. Lua触发事件-C#接收")]
    void TestLuaToCSharp()
    {
        luaInit.luaEnv.Global.Get<Action>("test_lua_to_csharp")?.Invoke();
    }
    
    [ContextMenu("4. C#触发事件-Lua接收")]
    void TestCSharpToLua()
    {
        Debug.Log("C# triggering event for Lua");
        luaCallback?.Invoke("C# to Lua message");
    }
    
    void OnDestroy()
    {
        luaCallback = null;
    }
    
    // 供Lua调用的方法
    public static void SendToCSharp(string message)
    {
        Debug.Log($"C# received from Lua: {message}");
    }
}

File: Assets/LearnXLua/LuaScripts/Lesson6_Event.lua
-- Lua事件表
local luaEventHandlers = {}

-- Lua触发事件，Lua接收
function test_lua_to_lua()
    print("Lua: Triggering event for Lua")
    for _, handler in ipairs(luaEventHandlers) do
        handler("Lua to Lua message")
    end
end

-- Lua触发事件，C#接收
function test_lua_to_csharp()
    print("Lua: Triggering event for C#")
    CS.Lesson6_Event.SendToCSharp("Lua to C# message")
end

-- Lua回调函数（供C#调用）
function on_lua_callback(message)
    print("Lua received from C#: " .. message)
end

-- 添加Lua事件监听器
function add_lua_event_listener(handler)
    table.insert(luaEventHandlers, handler)
end

-- 初始化时添加一个Lua事件监听器
add_lua_event_listener(function(message)
    print("Lua event handler: " .. message)
end)